plutus-scb/src/Cardano/Protocol/FollowerEffect.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE DataKinds        #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators    #-}
+
+module Cardano.Protocol.FollowerEffect where
+
+import           Control.Concurrent.STM
+import           Control.Lens                  hiding (index, use)
+import           Control.Monad.Freer
+import           Control.Monad.Freer.Extra.Log
+import           Control.Monad.Freer.Reader
+import           Control.Monad.Freer.State
+import           Control.Monad.IO.Class
+
+import           Data.List                     ((\\))
+
+import qualified Cardano.Node.Follower         as NF
+import qualified Cardano.Node.Types            as NT
+import           Ledger                        (Block)
+import qualified Ledger.Index                  as Index
+import           Wallet.Emulator.Chain         (ChainState (..), chainNewestFirst, currentSlot, index, txPool)
+
+{- | This handler will update it's local state whenever interpreting
+     a follower effect. -}
+handleNodeFollower
+  :: ( Member (State ChainState) effs
+     , Member (State NT.NodeFollowerState) effs
+     , Member (Reader (TQueue Block)) effs
+     , Member Log effs
+     , LastMember m effs
+     , MonadIO m
+     )
+  => Eff (NF.NodeFollowerEffect ': effs) ~> Eff effs
+handleNodeFollower =
+    (synchroniseState >>) . NF.handleNodeFollower
+
+synchroniseState
+  :: ( Member (State ChainState) effs
+     , Member (Reader (TQueue Block)) effs
+     , LastMember m effs
+     , MonadIO m
+     )
+  => Eff effs ()
+synchroniseState = fetchBlocks >>= injectBlocks
+
+fetchBlocks
+  :: ( Member (Reader (TQueue Block)) effs
+     , LastMember m effs
+     , MonadIO m
+     )
+  => Eff effs [Block]
+fetchBlocks =
+  ask >>= sendM . liftIO . atomically . flushTQueue
+
+injectBlocks
+  :: Member (State ChainState) effs
+  => [Block]
+  -> Eff effs ()
+injectBlocks = mapM_ (modify . inject)
+  where
+  {- The inject function is not very efficient, since it could
+     batch more blocks to update, but I think most of the time
+     we will only need to deal with one block updates -}
+  inject :: Block -> ChainState -> ChainState
```

Comment:
This function updates the `ChainState` with a new block. The other place where this happens is when a transaction gets validated (which performs more work than this does). I have extracted this into `addBlock` in `Chain.hs` (since it is about updating the chain). I am not sure if I should do anything more, like:

1) Add validation of the new block and stop assuming that the incoming block has been correctly validated by the Cardano node. If we do this, we should think about what to do with invalid blocks.

2) Use this function to update the state after the validation has been performed, but this will compute the utxo twice, once during validation and once during the state update.

Reply:
1. is different on the server and the client: the client should trust the server but the server does need to do validation.
I do see your point therefore that we end up with two functions. Not sure what to do about that. Computing the UTXO twice would almost certainly be fine for now, since the "efficient" version of the node will be the real one which will do something completely different.

Reply:
This change makes some tests fail. I will need more time to look into it. I think this changeset is large enough, I have added an issue to track these modifications: https://github.com/input-output-hk/plutus/issues/2140